Sources/Testing/Traits/Trait.swift

```diff
@@ -64,6 +64,45 @@ public protocol SuiteTrait: Trait {
   var isRecursive: Bool { get }
 }
 
+/// A protocol extending ``Trait`` that offers an additional customization point
+/// for trait authors to execute code before and after each test function (if
+/// added to the traits of a test function), or before and after each test suite
+/// (if added to the traits of a test suite).
+public protocol CustomExecutionTrait: Trait {
+
+  /// Execute a function with the effects of this trait applied.
+  ///
+  /// - Parameters:
+  ///   - function: The function to perform. If `test` represents a test suite,
+  ///     this function encapsulates running all the tests in that suite. If
+  ///     `test` represents a test function, this function is the body of that
+  ///     test function (including all cases if it is parameterized.)
+  ///   - test: The test under which `function` is being performed.
+  ///
+  /// - Throws: Whatever is thrown by `function`, or an error preventing the
+  ///   trait from running correctly.
+  ///
+  /// This method is called for each trait on a test suite or test function and
+  /// allows additional work to be performed before or after the test runs.
+  /// Issues recorded by this function are recorded against `test`.
+  ///
+  /// - Note: If a test function or test suite is skipped, this function does
+  ///   not get invoked by the runner.
+  ///
+  /// The default implementation of this function invokes `function` and
+  /// returns its result or rethrows any error it throws.
+  func execute(_ function: () async throws -> Void, for test: Test) async throws
```

Comment:
The API being proposed here will only be run once for each `Test`, and not once for each `Test.Case`, i.e. not once for every combination of arguments passed to a **parameterized** test.

I think it's very possible/likely some use cases will require performing the trait's custom execution logic once per test _case_, so we should probably figure out some way to support that via the API. I wonder if adding one additional parameter `testCase: Test.Case?` would accomplish that? (And adjusting the `Runner` implementation accordingly.)

Reply:
Adding a `Test.Case` parameter necessitates making this protocol SPI.

Reply:
#191 has now been merged, elevating `Test.Case` to public API (removing `@_spi`), so I think can now safely add a `testCase: Test.Case?` parameter to this method.